src/types/simple.rs

```diff
@@ -335,65 +353,78 @@ impl TypeRow {
     }
 
     /// Iterator over the types in the row.
-    pub fn iter(&self) -> impl Iterator<Item = &SimpleType> {
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
         self.types.iter()
     }
 
     /// Mutable iterator over the types in the row.
-    pub fn to_mut(&mut self) -> &mut Vec<SimpleType> {
+    pub fn to_mut(&mut self) -> &mut Vec<T> {
         self.types.to_mut()
     }
 
+    /// Allow access (consumption) of the contained elements
+    pub fn into_owned(self) -> Vec<T> {
+        self.types.into_owned()
+    }
+
     #[inline(always)]
     /// Returns the port type given an offset. Returns `None` if the offset is out of bounds.
-    pub fn get(&self, offset: usize) -> Option<&SimpleType> {
+    pub fn get(&self, offset: usize) -> Option<&T> {
         self.types.get(offset)
     }
 
     #[inline(always)]
     /// Returns the port type given an offset. Returns `None` if the offset is out of bounds.
-    pub fn get_mut(&mut self, offset: usize) -> Option<&mut SimpleType> {
+    pub fn get_mut(&mut self, offset: usize) -> Option<&mut T> {
         self.types.to_mut().get_mut(offset)
     }
 
-    #[inline]
-    /// Return the type row of variants required to define a Sum of Tuples type
-    /// given the rows of each tuple
-    pub fn predicate_variants_row(variant_rows: impl IntoIterator<Item = TypeRow>) -> Self {
-        variant_rows
+    fn try_convert_elems<D: PrimType + TryFrom<T>>(self) -> Result<TypeRow<D>, D::Error> {
+        let elems: Vec<D> = self
+            .into_owned()
             .into_iter()
-            .map(|row| SimpleType::Classic(ClassicType::Container(Container::Tuple(Box::new(row)))))
-            .collect_vec()
-            .into()
+            .map(D::try_from)
+            .collect::<Result<_, _>>()?;
+        Ok(TypeRow::from(elems))
+    }
+
+    /// Converts the elements of this TypeRow into some other type that they can `.into()`
+    pub fn map_into<T2: PrimType + From<T>>(self) -> TypeRow<T2> {
```

Comment:
Similarly, it'd be good for this just to be an Into/From impl, but again that seems to raise conflicting-instance problems